File: solutions/elixir/009-container-with-most-water.ex
# Problem 9: Container With Most Water (LeetCode 11)
# Difficulty: Med
# Language: Elixir
# 
defmodule ContainerWater do
  def max_area(height) do
    arr = :array.from_list(height)
    solve(arr, 0, :array.size(arr) - 1, 0)
  end

  defp solve(_arr, l, r, best) when l >= r, do: best
  defp solve(arr, l, r, best) do
    hl = :array.get(l, arr)
    hr = :array.get(r, arr)
    area = min(hl, hr) * (r - l)
    new_best = max(best, area)
    if hl < hr,
      do: solve(arr, l + 1, r, new_best),
      else: solve(arr, l, r - 1, new_best)
  end
end

File: solutions/ocaml/009-container-with-most-water.ml
(*
  Problem 9: Container With Most Water (LeetCode 11)
  Difficulty: Med
  Language: OCaml
*)
let max_area height =
  let arr = Array.of_list height in
  let n = Array.length arr in
  let rec solve l r best =
    if l >= r then best
    else
      let area = min arr.(l) arr.(r) * (r - l) in
      let new_best = max best area in
      if arr.(l) < arr.(r) then solve (l + 1) r new_best
      else solve l (r - 1) new_best
  in
  solve 0 (n - 1) 0
